server/src/middleware/json_middleware.rs

```diff
@@ -0,0 +1,75 @@
+use actix_http::header::HeaderValue;
+use actix_web::{
+    dev::{forward_ready, Service, ServiceRequest, ServiceResponse, Transform},
+    web::{self},
+    Error,
+};
+use futures_util::future::LocalBoxFuture;
+use std::{
+    future::{ready, Ready},
+    rc::Rc,
+};
+
+pub struct JsonMiddleware<S> {
+    service: Rc<S>,
+}
+
+impl<S, B> Service<ServiceRequest> for JsonMiddleware<S>
+where
+    S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error> + 'static,
+    S::Future: 'static,
+    B: 'static,
+{
+    type Response = ServiceResponse<B>;
+    type Error = Error;
+    type Future = LocalBoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    forward_ready!(service);
+    fn call(&self, mut req: ServiceRequest) -> Self::Future {
+        // Clone the Rc pointers so we can move them into the async block.
+        let srv = self.service.clone();
+        Box::pin(async move {
+            let bytes = req.extract::<web::Bytes>().await;
+            if let Ok(bytes) = bytes {
+                let json_data = serde_json::from_slice::<serde_json::Value>(&bytes);
+                if json_data.is_ok()
+                    && req
+                        .headers()
+                        .get(actix_http::header::CONTENT_TYPE)
+                        .is_none()
+                {
+                    req.headers_mut().insert(
+                        actix_http::header::CONTENT_TYPE,
+                        HeaderValue::from_static("application/json"),
+                    );
+                }
```

Comment:
No need to do check for if json payload is valid, JSON validator will automatically do that. Just need to set content type here

Reply:
The extractor uses the header to check before checking the body

Reply:
so we can't rely on the actix web extractor

Reply:
Yes, but it will just error back to the user if it is other than json. We just want to reduce any latency within the middleware. Specifically, having to deserialize the payload twice is wasteful